Added Liquid Minecarts

sparks0 opened the pull request:
Liquid Minecarts first draft. I threw this together in a few days so I'm sure there are errors or places where performance can be improved.

Video: https://www.youtube.com/watch?v=qQctcD3WvWw

Liquid Minecarts are used to move liquids between tanks. This module needs feedback and testing. The goal in its design is to support any new liquids automatically. Instead of hard-coded values, liquid minecarts copy the scores, tags and textures from the tank they load from.

Minecarts can hold 40% of the capacity of the tank from which they got their liquid from (usually 120 bottles).
Minecarts are created by placing down a hopper minecart and entering the same recipe as a liquid tank.
Minecarts will fill if there is a liquid tank pointed at them and they are empty or the liquid in the tank matches their own.
Minecarts will empty their contents into an empty tank or a tank with matching liquid if they are on top of the tank.
Minecarts will empty completely if they drive over a powered activator rail. This is so automated systems can assure tanks have no residual contents.
Minecarts have no GUI for survival players and have no functionality outside of being able to interface with tanks. Think of them as pipes.


A note on why liquid minecarts are command_block_minecart entities. Originally I used regular minecarts  but this caused the passenger tank armour stand to be ejected when going over an activator rail. Minecarts themselves don't support skulls as a custom display block. I didn't want the minecarts to have a GUI because they have no GUI functionality and seeing a hopper interface on a mobile liquid tank will cause players to expect it to be able to bottle and unbottle. Furthermore a GUI will cause the tank to suck items out of hoppers or place them in, potentially messing up auto-filler systems if it picks something up while travelling down a track. Furnace minecarts drive weird as we all know so command block minecart it is! There's an added advantage that the "drain" feature is entered as a command in the minecart itself meaning there's no 20hz clock needed to check for activator rails. As far as I can tell, in survival it acts like a normal minecart. You can't view the command GUI and breaking it doesn't drop a command block.

Kroppeb commented on gm4_liquid_minecarts/data/liquid_minecarts/functions/load_liquid.mcfunction:
```diff
@@ -0,0 +1,17 @@
+scoreboard players operation @s gm4_lm_data = @s gm4_lt_max
+scoreboard players operation @s gm4_lm_data -= @s gm4_lt_value
+#lower lm_data if tank has less liquid than the remaining space
+execute if score @s gm4_lm_data > @e[type=armor_stand,limit=1,sort=nearest,tag=gm4_liquid_tank] gm4_lt_value run scoreboard players operation @s gm4_lm_data = @e[type=armor_stand,limit=1,sort=nearest,tag=gm4_liquid_tank] gm4_lt_value
```
```suggestion
scoreboard players operation @s gm4_lm_data < @e[type=armor_stand,limit=1,sort=nearest,tag=gm4_liquid_tank] gm4_lt_value
```

Kroppeb commented on gm4_liquid_minecarts/data/liquid_minecarts/functions/minecart_to_tank_init.mcfunction:
```diff
@@ -0,0 +1,5 @@
+tag @s add gm4_liquid_tank
```
Missing header

Kroppeb commented on gm4_liquid_minecarts/data/liquid_minecarts/functions/create.mcfunction:
```diff
@@ -0,0 +1,9 @@
+execute as @e[type=command_block_minecart,distance=..0,tag=gm4_liquid_minecart_empty] run function liquid_minecarts:liquid_value_update
+scoreboard players set @e[type=command_block_minecart,distance=..0,tag=gm4_liquid_minecart_empty] gm4_lt_value 0
+scoreboard players set @e[type=command_block_minecart,distance=..0,tag=gm4_liquid_minecart_empty] gm4_lt_max 40
+scoreboard players set @e[type=command_block_minecart,distance=..0,tag=gm4_liquid_minecart_empty] gm4_lt_disp_val 0
```
`liquid_minecarts:liquid_value_update` will init the following values for `@s`
* `gm4_lt_value` = 0
* `gm4_lm_data` = 0
* `gm4_lt_max` = 0
* `gm4_lt_disp_val` = 0

These 3 initialisations are not needed. (Also why are you setting max to 40)

Kroppeb commented on gm4_liquid_minecarts/data/liquid_minecarts/functions/main.mcfunction:
```diff
@@ -0,0 +1,13 @@
+
```
Missing header

sparks0 replied:
Main doesn't need a header XD

Kroppeb replied:
Whoops

Kroppeb commented on gm4_liquid_minecarts/data/liquid_minecarts/functions/unload_liquid.mcfunction:
```diff
@@ -0,0 +1,19 @@
+scoreboard players operation @s gm4_lm_data = @e[type=armor_stand,limit=1,sort=nearest,tag=gm4_liquid_tank] gm4_lt_max
+scoreboard players operation @s gm4_lm_data -= @e[type=armor_stand,limit=1,sort=nearest,tag=gm4_liquid_tank] gm4_lt_value
+#lower lm_data if tank has less space than minecart's liquid level
+execute if score @s gm4_lt_value < @s gm4_lm_data run scoreboard players operation @s gm4_lm_data = @s gm4_lt_value
```
```suggestion
scoreboard players operation @s gm4_lm_data < @s gm4_lt_value
```

sparks0 commented:
Updated Liquid Minecarts

- Base on kroppeb's feedback I've made some performance improvements.
- Changed the display's pose so the hand item cannot be seen from outside the cart.
